### PR TITLE
fix mypy errors; add mypy precommit hooks

### LINKED ISSUE #4697: Add type-hints
There are lots of type hints throughout dvc's codebase, but I couldn't find any mention of which type checker the hints are expected to pass against.  I would propose mentioning it in the README and running the checker as part of the test suite.

### PR DESCRIPTION
**Pull request by skshetry**
Closes #4697
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/repo/add.py [hunk 1]
```diff
@@ -14,20 +15,29 @@
...
+
 logger = logging.getLogger(__name__)
 
 
 @locked
 @scm_context
 def add(
-    repo, targets, recursive=False, no_commit=False, fname=None, **kwargs,
+    repo,
+    targets: "TargetType",
```
Comment (skshetry) [inline dvc/repo/add.py:29]: I'd have used `TargetType` and `ensure_list` in more places, but most of the places need to leave `None` as-is to be used in `collect` or `used_cache`, so we need something like `ensure_list_or_none()`. Let me know if that makes sense, I'll add them in the next PR.

```python
def ensure_list_or_none(item: Union[List[str], str, None]) -> Optional[List[str]]:
    if item is None:
        return item
    if isinstance(item, str):
       return [item]
    return list(item)
```

**Review by pared — APPROVED** [review]

File: dvc/repo/add.py (hunk 1 above)
Comment (skshetry) [reply dvc/repo/add.py:29]: I am talking about the following stuff that are repeated on most of the API.
https://github.com/iterative/dvc/blob/ccc17917cfcb1b9ab8bc40d5ffdc8bcd02a1040a/dvc/repo/push.py#L22-L23

Comment (efiop) [reply dvc/repo/add.py:29]: A bit ugly, but makes sense.

**Review by efiop — APPROVED** [review]
Thank you @skshetry ! :pray:

**Pull request merged**